### DISCUSSION TITLE
Is it possible to build a node graph based software?

### DISCUSSION THREAD

**Discussion by sinkyl** [slint-ui/slint#3608] [Q&A] [answered]
Hi,

I want to build a node graph based software for data processing with Rust like this one (Unity/C#): [https://github.com/alelievr/NodeGraphProcessor](https://github.com/alelievr/NodeGraphProcessor)
Because you guys know Slint's limitations, I'm hoping to save time and energy.
If there are limitations can you name a few ones, if it's not possible or you think it will be complicated it's OK I'll look somewhere else.
Thank you in advance,

**Comment by ogoffart** [comment, marked as answer]
Hi,

I hope that this will be possible, but it may be some work.
The way to do it is to have the nodes and link and other elements in models and draw that.

Here is an example of very basic one
[SlintPad Example](https://slintpad.com/?snippet=struct+Node+%7B+x%3A+length%2C+y%3A+length%2C+text%3A+string+%7D%0Astruct+Link+%7B+x1%3A+length%2C+x2%3A+length%2C+y1%3A+length%2C+y2%3A+length+%7D%0A%0Aexport+component+Demo+inherits+Window+%7B%0A++++preferred-height%3A+500px%3B%0A++++preferred-width%3A+500px%3B%0A++++in-out+property+%3C%5BNode%5D%3E+nodes%3A+%5B%0A++++++++%7B+text%3A+%22Node1%22%2C+x%3A+120px%2C++y%3A+235px%2C+%7D+%2C%0A++++++++%7B+text%3A+%22Node2%22%2C+x%3A+350px%2C++y%3A+150px%2C+%7D+%2C%0A++++++++%7B+text%3A+%22Node3%22%2C+x%3A+300px%2C++y%3A+350px%2C+%7D+%2C%0A++++%5D%3B%0A%0A++++function+create-link%28a%3Aint%2C+b%3Aint%29-%3E+Link+%7B%0A++++++++return+%7B+x1%3A+nodes%5Ba%5D.x%2C+y1%3A+nodes%5Ba%5D.y%2C+x2%3A+nodes%5Bb%5D.x%2C+y2%3A+nodes%5Bb%5D.y+%7D%3B%0A++++%7D%0A++++in+property+%3C%5BLink%5D%3E+links%3A+%5B+create-link%281%2C2%29%2C+create-link%280%2C2%29+%5D%3B%0A%0A++++for+l+in+links+%3A+path+%3A%3D+Path+%7B%0A++++++++viewbox-width%3A+self.width+%2F1px%3B%0A++++++++viewbox-height%3A+self.height+%2F1px%3B%0A++++++++stroke%3A+orange%3B%0A++++++++stroke-width%3A+2px%3B%0A++++++++MoveTo+%7B+x%3A+l.x1+%2F+1px%3B+y%3A+l.y1+%2F+1px%3B+%7D%0A++++++++LineTo+%7B+x%3A+l.x2+%2F+1px%3B+y%3A+l.y2+%2F+1px%3B+%7D%0A%0A++++%7D%0A%0A++++for+n%5Bidx%5D+in+nodes+%3A+Rectangle+%7B%0A++++++++background%3A+gray%3B%0A++++++++x%3A+n.x+-+self.width%2F2%3B%0A++++++++y%3A+n.y+-+self.height%2F2%3B%0A++++++++width%3A+l.preferred-width%3B%0A++++++++height%3A+l.preferred-height%3B%0A++++++++border-radius%3A+3px%3B%0A++++++++l+%3A%3D+HorizontalLayout+%7B%0A++++++++++++padding%3A+15px%3B%0A++++++++++++Text+%7B+text%3A+n.text%3B+%7D%0A++++++++%7D%0A++++++++TouchArea+%7B%0A++++++++++++moved+%3D%3E+%7B%0A++++++++++++++++if+%28self.pressed%29+%7B%0A++++++++++++++++++++n.x+%2B%3D+self.mouse-x+-+self.pressed-x%3B%0A++++++++++++++++++++n.y+%2B%3D+self.mouse-y+-+self.pressed-y%3B%0A++++++++++++++++++++nodes%5Bidx%5D+%3D+n%3B%0A++++++++++++++++%7D%0A++++++++++++%7D%0A++++++++++++mouse-cursor%3A+move%3B%0A++++++++%7D%0A++++%7D%0A%7D)
But of course, the exact logic should be done in native code and not in Slint.

**Reply by sinkyl** [reply]
I want to thank you for your quick response and for the example you provided.
I was thinking of using Rust for the business logic and Slint for the UI.
So I'll give it a try and see if it fits the requirements.
Thank you again.
Best,

**Reply by rizary** [reply]
@sinkyl did you tried?
